File: DSA/Implementations/StackArray.java
import java.util.EmptyStackException;

/**
 * A implementation of a stack using an dynamically resizable array.
 * It supports typical stack operations:
 * {@code push}, {@code pop}, {@code peek}, {@code isEmpty}, {@code size}.
 * The internal array grows and shrinks as needed
 *
 * @author Aniket
 * @version 1.0
 */

public class StackArray {
  private int[] array;
  private int size;
  private int capacity;
  
  /**
   * Crates an empty stack with initial capacity of 1.
   */
  public StackArray() {
    array = new int[1];
    capacity = 1;
    size = 0;
  } 
 
  /**
   * Resizes the array to a new capacity.
   *
   * @param newCapacity the new capacity for the array
   */
  public void resize(int newCapacity) {
    int[] newArray = new int[newCapacity];
    for (int i = 0; i < size; i++) {
      newArray[i] = array[i];
    }
    array = newArray;
  }

  /**
   * Pushes a value to the top of the stack.
   * If the array is full, its capacity is doubled.
   *
   * @param value the value to be pushed
   */
  public void push(int value) {
    if (size == capacity) {
      capacity *= 2;
      resize(capacity);
    }
    array[size++] = value;
  }

  /** Removes and returns the top element of the stack.
 *
   * @return the value at the top of the stack
   * @throws EmptyStackException if the stack is empty
   */
  public int pop() {
    if (isEmpty()) {
      throw new EmptyStackException();
    }
    int item = array[--size];
    if (size > 0 && size <= capacity / 4) {
      capacity /= 2;
      resize(capacity);
    }
    return item;
  }

  /** Returns the value at the top of the stack.
   *
   * @return the value at the top of the stack
   * @throws EmptyStackException if the stack is empty
   */
  public int peek() {
    if (isEmpty()) {
      throw new EmptyStackException();
    }
    return array[size - 1];
  }

  /** Checks if the stack is empty.
   *
   * @return {@code true} if the stack is empty; {@false} otherwise
   */
  public boolean isEmpty() {
    return size == 0;
  }

  /** Returns the number of element in the stack.
   *
   * @return the size of the stack
   */
  public int size() {
    return size;
  }
  
  /** Runs basic tests for the stack implementaion.
   *
   * @param args unused
   */
  public static void main(String[] args) {
    StackArray stack = new StackArray();

    stack.push(3);
    stack.push(5);
    stack.push(7);

    assert stack.size() == 3;
    assert !stack.isEmpty();

    assert stack.peek() == 7;
    assert stack.pop() == 7;
    assert stack.size() == 2;
    assert !stack.isEmpty();

    assert stack.peek() == 5;
    assert stack.pop() == 5;
    assert stack.size() == 1;
    assert !stack.isEmpty();

    assert stack.peek() == 3;
    assert stack.pop() == 3;
    assert stack.size() == 0;
    assert stack.isEmpty();

    System.out.println("All stack tests passed!");
  }
}

File: DSA/Implementations/StackArray.c
// Copyright 2025 Aniket

#include <assert.h>
#include <limits.h>
#include <stdbool.h>
#include <stdio.h>
#include <stdlib.h>

void push(int value);
int pop();
int peek();
bool isEmpty();
int size();
void resize_array(int new_capacity);

int *array;
int N = 0;
int max_size = 1;

void resize_array(int new_capacity) {
  int *new_array = (int *)malloc(sizeof(int) * new_capacity);

  for (int i = 0; i < N; i++) {
    new_array[i] = array[i];
  }
  free(array);
  array = new_array;
  max_size = new_capacity;
}

void push(int item) {
  if (N >= max_size) {
    resize_array(max_size * 2);
  }
  array[N++] = item;
}

int peek() {
  if (isEmpty()) {
    fprintf(stderr, "Error: empty stack\n");
    exit(EXIT_FAILURE);
  }
  return array[N - 1];
}

int pop() {
  if (isEmpty()) {
    fprintf(stderr, "Error: Stack Underflow\n");
    exit(EXIT_FAILURE);
  }
  int item = array[--N];
  if (N > 0 && N <= max_size / 4) {
    max_size /= 2;
    if (max_size == 0) max_size = 1;
    resize_array(max_size);
  }
  return item;
}

bool isEmpty() { return N == 0; }

int size() { return N; }

void testStack() {
  push(10);
  push(20);
  push(30);

  assert(size() == 3);
  assert(isEmpty() == false);

  assert(peek() == 30);
  assert(size() == 3);

  assert(pop() == 30);
  assert(size() == 2);
  assert(peek() == 20);

  assert(pop() == 20);
  assert(size() == 1);
  assert(peek() == 10);

  assert(pop() == 10);
  assert(size() == 0);
  assert(isEmpty() == true);

  printf("All stack tests passed!\n");
}

int main() {
  array = (int *)malloc(sizeof(int) * max_size);
  testStack();
  free(array);
  return 0;
}
